# 📚
# **STUDENT ACADEMIC RISK PREDICTION** 
## **Notebook 01: Data Acquisition & Validation**
**Author:** Ahossey Paul  
**Date:** 2025  
**Dataset:** UCI Student Performance Dataset  
**Goal:** Load, inspect, and validate raw data before any processing

In [3]:
# Standard libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)
sns.set_theme(style='whitegrid', palette='muted')

print("Imports successful!")

Imports successful!


In [5]:
# Load the Data:
# Define paths
RAW_DATA_PATH = '../data/raw/'

df_math = pd.read_csv(RAW_DATA_PATH + 'student-mat.csv', sep=';')
df_port = pd.read_csv(RAW_DATA_PATH + 'student-por.csv', sep=';')

print(f"Math dataset shape: {df_math.shape}")
print(f"Portuguese dataset shape: {df_port.shape}")

Math dataset shape: (395, 33)
Portuguese dataset shape: (649, 33)


In [7]:
#First Inspection
print(f"Rows: {df_math.shape[0]} | Columns: {df_math.shape[1]}")
print(f"\nColumn names:\n{list(df_math.columns)}")
print(f"\nFirst 10 rows:")

df_math.head(10)

Rows: 395 | Columns: 33

Column names:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2', 'G3']

First 10 rows:


,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,reason,guardian,traveltime,studytime,failures,schoolsup,famsup,paid,activities,nursery,higher,internet,romantic,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
0,GP,F,18,U,GT3,A,4,4,at_home,teacher,course,mother,2,2,0,yes,no,no,no,yes,yes,no,no,4,3,4,1,1,3,6,5,6,6
1,GP,F,17,U,GT3,T,1,1,at_home,other,course,father,1,2,0,no,yes,no,no,no,yes,yes,no,5,3,3,1,1,3,4,5,5,6
2,GP,F,15,U,LE3,T,1,1,at_home,other,other,mother,1,2,3,yes,no,yes,no,yes,yes,yes,no,4,3,2,2,3,3,10,7,8,10
3,GP,F,15,U,GT3,T,4,2,health,services,home,mother,1,3,0,no,yes,yes,yes,yes,yes,yes,yes,3,2,2,1,1,5,2,15,14,15
4,GP,F,16,U,GT3,T,3,3,other,other,home,father,1,2,0,no,yes,yes,no,yes,yes,no,no,4,3,2,1,2,5,4,6,10,10
5,GP,M,16,U,LE3,T,4,3,services,other,reputation,mother,1,2,0,no,yes,yes,yes,yes,yes,yes,no,5,4,2,1,2,5,10,15,15,15
6,GP,M,16,U,LE3,T,2,2,other,other,home,mother,1,2,0,no,no,no,no,yes,yes,yes,no,4,4,4,1,1,3,0,12,12,11
7,GP,F,17,U,GT3,A,4,4,other,teacher,home,mother,2,2,0,yes,yes,no,no,yes,yes,no,no,4,1,4,1,1,1,6,6,5,6
8,GP,M,15,U,LE3,A,3,2,services,other,home,mother,1,2,0,no,yes,yes,no,yes,yes,yes,no,4,2,2,1,1,1,0,16,18,19
9,GP,M,15,U,GT3,T,3,4,other,other,home,mother,1,2,0,no,yes,yes,yes,yes,yes,yes,no,5,5,1,1,1,5,0,14,15,15


In [10]:
# Data Quality Check
def data_quality_report(df, name):
    """
    Generate a professional data quality summary.
    Every real project starts with this.
    """
    
    print(f"\033[1mDATA QUALITY REPORT: {name}\033[0m")
    
    print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")
    print(f"\nMissing Values: {df.isnull().sum().sum()} total")
    print(f"Duplicate Rows: {df.duplicated().sum()}")
    print(f"\nData Types:")
    print(df.dtypes.value_counts())
    print(f"\nNumeric Summary:")
    return df.describe()

data_quality_report(df_math, "Student Math Performance")

DATA QUALITY REPORT: Student Math Performance
Shape: 395 rows × 33 columns

Missing Values: 0 total
Duplicate Rows: 0

Data Types:
object    17
int64     16
Name: count, dtype: int64

Numeric Summary:


,age,Medu,Fedu,traveltime,studytime,failures,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
count,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00,395.00
mean,16.70,2.75,2.52,1.45,2.04,0.33,3.94,3.24,3.11,1.48,2.29,3.55,5.71,10.91,10.71,10.42
std,1.28,1.09,1.09,0.70,0.84,0.74,0.90,1.00,1.11,0.89,1.29,1.39,8.00,3.32,3.76,4.58
min,15.00,0.00,0.00,1.00,1.00,0.00,1.00,1.00,1.00,1.00,1.00,1.00,0.00,3.00,0.00,0.00
25%,16.00,2.00,2.00,1.00,1.00,0.00,4.00,3.00,2.00,1.00,1.00,3.00,0.00,8.00,9.00,8.00
50%,17.00,3.00,2.00,1.00,2.00,0.00,4.00,3.00,3.00,1.00,2.00,4.00,4.00,11.00,11.00,11.00
75%,18.00,4.00,3.00,2.00,2.00,0.00,5.00,4.00,4.00,2.00,3.00,5.00,8.00,13.00,13.00,14.00
max,22.00,4.00,4.00,4.00,4.00,3.00,5.00,5.00,5.00,5.00,5.00,5.00,75.00,19.00,19.00,20.00


In [11]:
print(df_math['G3'].describe())
print(f"\nUnique values in G3: {sorted(df_math['G3'].unique())}")

count   395.00
mean     10.42
std       4.58
min       0.00
25%       8.00
50%      11.00
75%      14.00
max      20.00
Name: G3, dtype: float64

Unique values in G3: [0, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


In [12]:
# How many students are at risk?
threshold = 10

df_math['at_risk'] = (df_math['G3'] < threshold).astype(int)

total = len(df_math)
at_risk_count = df_math['at_risk'].sum()
passing_count = total - at_risk_count

print(f"Total students:     {total}")
print(f"At risk (fail):     {at_risk_count} ({at_risk_count/total*100:.1f}%)")
print(f"Passing:            {passing_count} ({passing_count/total*100:.1f}%)")

Total students:     395
At risk (fail):     130 (32.9%)
Passing:            265 (67.1%)


In [13]:
#Combine Both Subjects

# Add subject label before merging
df_math['subject'] = 'math'
df_port['subject'] = 'portuguese'

# Combine into one master dataset
df = pd.concat([df_math, df_port], ignore_index=True)

print(f"Combined dataset shape: {df.shape}")
print(f"Subject breakdown:\n{df['subject'].value_counts()}")

Combined dataset shape: (1044, 35)
Subject breakdown:
subject
portuguese    649
math          395
Name: count, dtype: int64


In [14]:
# Create Target Variable

# Binary target: 1 = at risk (failing), 0 = passing
df['at_risk'] = (df['G3'] < 10).astype(int)

print(f"Class distribution:\n{df['at_risk'].value_counts()}")
print(f"\nAt-risk rate: {df['at_risk'].mean()*100:.1f}%")

Class distribution:
at_risk
0    814
1    230
Name: count, dtype: int64

At-risk rate: 22.0%


In [16]:
#Save to Processed Folder
# Save combined dataset — this becomes our working file going forward
PROCESSED_PATH = '../data/processed/'

df.to_csv(PROCESSED_PATH + 'student_combined.csv', index=False)

print(f"Dataset saved successfully to {PROCESSED_PATH}student_combined.csv")
print(f"Final shape: {df.shape}")

Dataset saved successfully to ../data/processed/student_combined.csv
Final shape: (1044, 35)


## Data Sources

| Item | Detail |
|---|---|
| **Source** | UCI Machine Learning Repository |
| **Dataset** | Student Performance Dataset |
| **Citation** | Cortez & Silva, 2008 |
| **Subjects** | Mathematics + Portuguese Language |
| **Total Records** | 1,044 students |
| **Features** | 33 original + 2 engineered (subject, at_risk) |
| **Target** | at_risk (binary: G3 < 10) |
| **License** | Public, research use permitted |